### ISSUE TITLE
Address pandas deprecations

### ISSUE THREAD

**Issue by ekatef** [pz-max/pypsa-africa#687]
The recent [pandas 2.0 release](https://pandas.pydata.org/docs/dev/whatsnew/v2.0.0.html) at 3 April has removed append for both Series and Dataframes. There is a need to check where pandas and geopandas `append` are used in the code

Thanks a lot to @yerbol-akhmetov for reporting!

## Checklist

- [х] I am using the current [`main`](https://github.com/pypsa-meets-earth/pypsa-earth/tree/main) branch or the latest [release](https://github.com/pypsa-meets-earth/pypsa-earth/releases). Please indicate.
- [х] I am running on an up-to-date [`pypsa-earth` environment](https://github.com/pypsa-meets-earth/pypsa-earth/blob/main/envs/environment.yaml). Update via `conda env update -f envs/environment.yaml`.

## Describe the Bug

Any attempt to call pandas `append` results in an error `AttributeError: 'DataFrame' object has no attribute 'append'`, if an environment is using pandas `version>2.0.0`.

## Error Message

Currently, one of the ways to reproduce an error is to run a workflow with `custom_powerplants: true` (will be fixed with #686):

```bash
Traceback (most recent call last):
  File "~/pypsa-earth/.snakemake/scripts/tmpl0i66xpm.build_powerplants.py", line 314, in <module>
    ppl = add_custom_powerplants(
  File "~/pypsa-earth/.snakemake/scripts/tmpl0i66xpm.build_powerplants.py", line 254, in add_custom_powerplants
    return ppl.append(add_ppls, sort=False, ignore_index=True, verify_integrity=True)
  File "~/miniconda3/envs/pypsa-earth-upd6/lib/python3.10/site-packages/pandas/core/generic.py", line 5989, in __getattr__
    return object.__getattribute__(self, name)
AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
```

**Comment by ekatef** [comment]
NB using list `append` is perfectly alright. Actually, the reason why pandas `append` has been removed was avoiding confusion between list and pandas `append` methods

**Comment by ekatef** [comment]
Quick scan of the workflow with tutorial (more or less standard) settings don't throw any problems. So, hopefully the issue is rather minor

**Comment by ekatef** [comment]
Apart of `append` deprecation, it may be also good to fix other pandas deprecations, like `'axis' keyword in DataFrame.groupby`.

**Comment by ekatef** [comment]
The list of revisions needed seems to be not very long. Most of the deprecation warnings are caused by using an old PyPSA version. Purely panda-related deprecations are as follows:

- [ ] **clean_osm_data**

- `Setting an item of incompatible dtype is deprecated` in `df.loc[filled_values.index, "circuits"] = filled_values`

- `DataFrame.applymap has been deprecated. Use DataFrame.map instead.` in `is_all_list = df[cols].applymap(lambda x: isinstance(x, list)).all(axis=1)` and `nunique_values = df[cols].applymap(len).nunique(axis=1)`

- [ ] **build_osm_network**

`FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated` in `pd.concat([lines, df_to_add], ignore_index=True).reset_index(drop=True)`

- [ ] **prepare_network**

`Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated` in `df = df.loc[:, "Y_1970":"Y_2018"].ffill(axis=1)`

**Comment by davide-f** [comment]
Great @ekatef :D
Completely agree that this is needed; for coordination, are you working on that or will you be planning to do so?

**Comment by ekatef** [comment]
> Great @ekatef :D Completely agree that this is needed; for coordination, are you working on that or will you be planning to do so?

Thanks for coordination @davide-f! 😄 No, I haven't started to work on that, and contributions are very welcome.

Not sure when the next major `pandas` release is to be expected: [in the docs it looks like](https://pandas.pydata.org/docs/dev/whatsnew/v3.0.0.html) that may be the case in a few months.

**Comment by finozzifa** [comment]
Hi all,
I am happy to take on this issue if you like!

**Comment by ekatef** [comment]
> Hi all, I am happy to take on this issue if you like!

Hey @finozzifa, it would be immensely appreciated! 😄

**Comment by finozzifa** [comment]
ok. That's mine :D

**Comment by davide-f** [comment]
Greatly greatly appreciated :D

**Comment by finozzifa** [comment]
Hi @ekatef and @davide-f 

The deprecations to be addressed are:
 
- [x] **add_electricity**:  /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmptfj309dw.add_electricity.py:332: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy. For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object. df.carrier.mask(df.technology == "Onshore", "onwind", inplace=True)

- [x] **clean_osm_data**: Setting an item of incompatible dtype is deprecated in df.loc[filled_values.index, "circuits"] = filled_values
- [x] **clean_osm_data**: DataFrame.applymap has been deprecated. Use DataFrame.map instead. in is_all_list = df[cols].applymap(lambda x: isinstance(x, list)).all(axis=1) and nunique_values = df[cols].applymap(len).nunique(axis=1)

- [x] **build_osm_network**: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation. pd.concat([lines, df_to_add], ignore_index=True).reset_index(drop=True)

- [x] **build_osm_network**:  /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmprqe2v9pu.build_osm_network.py:178: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like: df["col"][row_indexer] = value Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

- [x] **prepare_network**: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated in df = df.loc[:, "Y_1970":"Y_2018"].ffill(axis=1)

- [x] **prepare_network**: /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmp3cpcw0gx.prepare_network.py:312: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method. The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy. For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object. n.lines.s_nom_max.clip(upper=s_nom_max_set, inplace=True)

- [x] **prepare_network**: /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmp3cpcw0gx.prepare_network.py:313: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy. For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object. n.links.p_nom_max.clip(upper=p_nom_max_set, inplace=True)

I have also found the following ones. Should I address them as well? :
- [x] **cluster_network**: ./home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmpbb_k9znc.cluster_network.py:557: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning. .apply(busmap_for_country)

- [x] **prepare_network**:  /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmp3cpcw0gx.prepare_network.py:243: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead. pnl[k] = df.resample(offset).mean()


Thanks

**Comment by ekatef** [comment]
> Hi @ekatef and @davide-f
> 
> The deprecations to be addressed are:
> 
> * [x]  **add_electricity**:  /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmptfj309dw.add_electricity.py:332: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
>   The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy. For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object. df.carrier.mask(df.technology == "Onshore", "onwind", inplace=True)
> * [x]  **clean_osm_data**: Setting an item of incompatible dtype is deprecated in df.loc[filled_values.index, "circuits"] = filled_values
> * [x]  **clean_osm_data**: DataFrame.applymap has been deprecated. Use DataFrame.map instead. in is_all_list = df[cols].applymap(lambda x: isinstance(x, list)).all(axis=1) and nunique_values = df[cols].applymap(len).nunique(axis=1)
> * [x]  **build_osm_network**: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation. pd.concat([lines, df_to_add], ignore_index=True).reset_index(drop=True)
> * [x]  **build_osm_network**:  /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmprqe2v9pu.build_osm_network.py:178: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
>   You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
>   A typical example is when you are setting values in a column of a DataFrame, like: df["col"][row_indexer] = value Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.
> * [x]  **prepare_network**: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated in df = df.loc[:, "Y_1970":"Y_2018"].ffill(axis=1)
> * [x]  **prepare_network**: /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmp3cpcw0gx.prepare_network.py:312: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method. The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy. For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object. n.lines.s_nom_max.clip(upper=s_nom_max_set, inplace=True)
> * [x]  **prepare_network**: /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmp3cpcw0gx.prepare_network.py:313: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
>   The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy. For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object. n.links.p_nom_max.clip(upper=p_nom_max_set, inplace=True)
> 
> I have also found the following ones. Should I address them as well? : -[ ] **cluster_network**: ./home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmpbb_k9znc.cluster_network.py:557: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning. .apply(busmap_for_country)
> 
> -[ ] **prepare_network**: /home/runner/work/pypsa-earth/pypsa-earth/.snakemake/scripts/tmp3cpcw0gx.prepare_network.py:243: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead. pnl[k] = df.resample(offset).mean()
> 
> Thanks

Perfect, @finozzifa! Thanks a lot for the detailed tracking.

Absolutely agree with you that it makes absolute sense to address the additional deprecation issues you identified.

**Comment by ekatef** [comment]
Hello @finozzifa! Fantastic and much wanted work. The logs look now in fact way cleaner and we are ready to the future :D

Have added a couple of minor comments on interfaces of the time aggregation. Could you please let me know what do you think about them?

All the other fixes look perfect for me and basically ready to merge. @davide-f would you have time for the final review?

**Comment by finozzifa** [comment]
Hi Katia thanks again. I have included your suggestions and the pipeline is running. Thanks :)

**Comment by ekatef** [comment]
Closed as completed with #1023
